sklearn/metrics/cluster/tests/test_common.py

```diff
@@ -232,3 +239,88 @@ def test_returned_value_consistency(name):
+
+    a_xp = xp.asarray(a_np, device=device)
+    b_xp = xp.asarray(b_np, device=device)
+
+    metric_np = metric(a_np, b_np, **metric_kwargs)
+
+    # When array API dispatch is disabled, and np.asarray works (for example PyTorch
+    # with CPU device), calling the metric function with such numpy compatible inputs
+    # should work (albeit by implicitly converting to numpy arrays instead of
+    # dispatching to the array library).
+    try:
+        np.asarray(a_xp)
+        np.asarray(b_xp)
+        numpy_as_array_works = True
+    except (TypeError, RuntimeError, ValueError):
+        # PyTorch with CUDA device and CuPy raise TypeError consistently.
+        # array-api-strict chose to raise RuntimeError instead. NumPy raises
+        # a ValueError if the `__array__` dunder does not return an array.
+        # Exception type may need to be updated in the future for other libraries.
+        numpy_as_array_works = False
+
+    def _check_metric_matches(metric_a, metric_b, convert_a=False):
+        if convert_a:
+            metric_a = _convert_to_numpy(xp.asarray(metric_a), xp)
+        assert_allclose(metric_a, metric_b, atol=_atol_for_type(dtype_name))
+
+    if numpy_as_array_works:
+        metric_xp = metric(a_xp, b_xp, **metric_kwargs)
+
+        # Handle cases where multiple return values are not of the same shape,
+        # e.g. precision_recall_curve:
+        _check_metric_matches(metric_xp, metric_np)
+
+        metric_xp_mixed_1 = metric(a_np, b_xp, **metric_kwargs)
+        _check_metric_matches(metric_xp_mixed_1, metric_np)
+
+        metric_xp_mixed_2 = metric(a_xp, b_np, **metric_kwargs)
+        _check_metric_matches(metric_xp_mixed_2, metric_np)
+
+    with config_context(array_api_dispatch=True):
+        metric_xp = metric(a_xp, b_xp, **metric_kwargs)
+        _check_metric_matches(metric_xp, metric_np, convert_a=True)
+
+
+def check_array_api_unsupervised_metric(metric, array_namespace, device, dtype_name):
+    y_pred = np.array([1, 0, 1, 0, 1, 1, 0])
+    X = np.random.randint(10, size=(7, 10))
+
+    check_array_api_metric(
+        metric,
+        array_namespace,
+        device,
+        dtype_name,
+        a_np=X,
+        b_np=y_pred,
+    )
+
+
+array_api_metric_checkers = {
+    calinski_harabasz_score: [
+        check_array_api_unsupervised_metric,
+    ]
+}
+
+
+def yield_metric_checker_combinations(metric_checkers=array_api_metric_checkers):
+    for metric, checkers in metric_checkers.items():
+        for checker in checkers:
+            yield metric, checker
+
+
+@pytest.mark.parametrize(
+    "array_namespace, device, dtype_name",
+    yield_namespace_device_dtype_combinations(),
+    ids=_get_namespace_device_dtype_ids,
+)
+@pytest.mark.parametrize("metric, check_func", yield_metric_checker_combinations())
+def test_array_api_compliance(metric, array_namespace, device, dtype_name, check_func):
+    check_func(metric, array_namespace, device, dtype_name)
```

Comment:
I copied this logic (and modified it slightly) from `sklearn/metrics/tests/test_common.py`. 

If other cluster algorithms become array API compliant, this should make it simpler to test.